bigtop-packages/src/charm/zookeeper/layer-zookeeper/reactive/zookeeper.py

```diff
@@ -52,36 +66,190 @@ def update_network_interface():
-    Checkup on the state of the cluster. Inform an operator that they
-    need to restart if the peers have changed.
+    return [node[1].split(':')[0] for node in nodes]
+
+
+@when('zookeeper.started', 'leadership.is_leader', 'zkpeer.joined')
+@when_not('zkpeer.departed')
+def check_cluster(zkpeer):
+    '''
+    Checkup on the state of the cluster. Start a rolling restart if
+    the peers have changed.
 
     '''
     zk = Zookeeper()
     if data_changed('zkpeer.nodes', zk.read_peers()):
-        if zk.is_zk_leader():
-            note = ' (restart this node last)'
+        peers = _ip_list(zk.sort_peers(zkpeer))
+        nonce = time.time()
+        hookenv.log('Quorum changed. Restart queue: {}'.format(peers))
+        leader_set(
+            restart_queue=json.dumps(peers),
+            restart_nonce=json.dumps(nonce)
+        )
+
+
+@when('zookeeper.started', 'leadership.is_leader', 'zkpeer.joined',
+      'zkpeer.departed')
+def check_cluster_departed(zkpeer, zkpeer_departed):
+    '''
+    Wrapper around check_cluster.
+
+    Together with check_cluster, implements the following logic:
+
+    "Run this when zkpeer.joined and zkpeer departed, or zkpeer.joined
+    and not zkpeer.departed"
+
+    '''
+    check_cluster(zkpeer)
+
+
+@when('leadership.changed.restart_queue', 'zkpeer.joined')
+def restart_for_quorum(zkpeer):
+    '''
+    If we're the next node in the restart queue, restart, and then
+    inform the leader that we've restarted. (If we are the leader,
+    remove ourselves from the queue, and update the leadership data.)
+
+    '''
+    private_address = hookenv.unit_get('private-address')
+    queue = json.loads(leader_get('restart_queue'))
+
+    if not queue:
+        # Everything has restarted.
+        return
+
+    if private_address == queue[0]:
+        # It's our turn to restart.
+        _restart_zookeeper('updating_quorum')
+        if is_state('leadership.is_leader'):
+            queue = queue[1:]
+            hookenv.log('Leader updating restart queue: {}'.format(queue))
+            leader_set(restart_queue=json.dumps(queue))
         else:
-            note = ''
-        message = (
-            "number of zk peers has changed -- you must use "
-            "the 'restart' action to perform a rolling restart to "
-            "update your cluster{}".format(note))
-        hookenv.status_set('active', message)
+            zkpeer.inform_restart()
 
 
-@when('zookeeper.started', 'zookeeper.joined')
-def serve_client(client):
-    config = Zookeeper().dist_config
-    port = config.port('zookeeper')
-    rest_port = config.port('zookeeper-rest')  # TODO: add zookeeper REST
-    client.send_port(port, rest_port)
+@when('leadership.is_leader', 'zkpeer.restarted')
+def update_restart_queue(zkpeer):
+    '''
+    If a Zookeeper node has restarted as part of a rolling restart,
+    pop it off of the queue.
+
+    '''
+    queue = json.loads(leader_get('restart_queue'))[1:]
+    remove_state('zkpeer.restarted')
```

Comment:
Charm layers should never directly add or remove states set by an interface layer, as that breaks the encapsulation.  Additionally, it's conceivable that a restarted unit might trigger another `-changed` hook for some unrelated reason which would cause the `zkpeer.restarted` state to be re-added and this code to erroneously assume that some new unit has restarted.

Perhaps trying to track the units which have restarted with a state is the wrong approach.  What about something like this instead:

```python
@when('zookeeper.rolling_restart.in_progress', 'leadership.is_leader', 'zkpeer.joined')
def update_restart_queue(zkpeer):
    queue = json.loads(leader_get('restart_queue'))
    changed = False
    for peer in zkpeer.restarted_nodes():
        if peer in queue:
            queue.remove(peer)
            changed = True
    if changed:
        hookenv.log('Leader updating restart queue: {}'.format(queue))
        leader_set(restart_queue=json.dumps(queue))
```